typing/typemod.ml

```diff
@@ -241,21 +241,28 @@ let check_recmod_typedecls env decls =
 
 (* Merge one "with" constraint in a signature *)
 
-let rec add_rec_types env = function
-    Sig_type(id, decl, Trec_next, _) :: rem ->
-      add_rec_types (Env.add_type ~check:true id decl env) rem
-  | _ -> env
-
-let check_type_decl env loc id row_id newdecl decl rs rem =
-  let env = Env.add_type ~check:true id newdecl env in
-  let env =
+let check_type_decl env loc id row_id newdecl decl =
+  let fresh_id = Ident.rename id in
+  let path = Pident fresh_id in
+  let sub = Subst.add_type id path Subst.identity in
+  let fresh_row_id, sub =
     match row_id with
+    | None -> None, sub
+    | Some id ->
+      let fresh_row_id = Some (Ident.rename id) in
+      let sub = Subst.add_type id (Pident fresh_id) sub in
+      fresh_row_id, sub
+  in
+  let newdecl = Subst.type_declaration sub newdecl in
+  let decl = Subst.type_declaration sub decl in
+  let env = Env.add_type ~check:false fresh_id newdecl env in
+  let env =
+    match fresh_row_id with
     | None -> env
-    | Some id -> Env.add_type ~check:false id newdecl env
+    | Some fresh_row_id -> Env.add_type ~check:false fresh_row_id newdecl env
   in
-  let env = if rs = Trec_not then env else add_rec_types env rem in
-  Includemod.type_declarations ~mark:Mark_both ~loc env id newdecl decl;
-  Typedecl.check_coherence env loc (Path.Pident id) newdecl
+  Includemod.type_declarations ~mark:Mark_both ~loc env fresh_id newdecl decl;
+  Typedecl.check_coherence env loc path newdecl
```

Comment:
If I understand correctly, you are not considering the items that are part of the same recursive group anymore (this is what `add_rec_types` was doing, again iirc). Can you explain the change?

This looks a bit strange to me. I would think of a situation where `t` and `u` are mutually recursive, then `t` gets refined with a constraint; before the code would add both new-t and `u` and check them both, now you are not adding `u` anymore, the still definition is still in the environment but it itself refers to the old identifier for `t`, so checking it is always going to succeed. Instead I would expect that you would apply the name-freshening substitution to the other items of the recursive group (and I guess also freshen them?), and check them as a new group.

This being said, I can't think of a situation where not re-checking the other recursive declarations leads to a soundness issue. (But then I'm not very imaginative with this.) We are, after all, checking that the new declaration is included in the previous one, so morally this should ensure that old uses of the type, including in mutual definitions, are still correct?

Reply:
My example above was supposed to demonstrate that it isn't safe to just treat references to the old `t` as references to the new `t` whilst still checking that the old `t` and new `t` are compatible. It ended up showing that the problem is more fundamental than I thought, and the approach to recursion needs to be weakened further to avoid such issues.

At the same time, the approach to recursion in this PR is also too weak. It is sound, but too conservative. For example, it breaks the following example:
```ocaml
# module type S = sig
    type t = Foo of r
    and r = t
  end;;
      module type S = sig type t = Foo of r and r = t end
# type s = Foo of s;;
type s = Foo of s
# module type T = S with type t = s;;
Line 1, characters 23-33:
1 | module type T = S with type t = s;;
                           ^^^^^^^^^^
Error: This variant or record definition does not match that of type s
       Constructors do not match:
         Foo of s
       is not compatible with:
         Foo of r
       The types are not equal.
```

I'm now working on a new PR that should fix both these issues by making a more involved change to how we check `with type`.

Reply:
I should also point out that the previous code was definitely wrong. You cannot use `Trec_next` to reliably detect recursion in a signature -- it should only ever be used to improve printing never for actual type-checking.

At the moment, any example using recursion through modules fails
```ocaml
# module type S = sig
    module rec M : sig
      type t = Foo of M.r
      type r = t
    end
  end;;
module type S = sig module rec M : sig type t = Foo of M.r type r = t end end

# type s = Foo of s;;
type s = Foo of s

# module type T = S with type M.t = s;;
Line 1, characters 23-35:
1 | module type T = S with type M.t = s;;
                           ^^^^^^^^^^^^
Error: This variant or record definition does not match that of type s
       Constructors do not match:
         Foo of s
       is not compatible with:
         Foo of M.r
       The types are not equal.
```

Reply:
The behaviour for recursive groups is no longer changed.